### PR TITLE
Feature/can can can in members module

### PR DESCRIPTION
**Pull request by abdullah-siddiqui-git**
CANCANCAN is integration into Members module. Moreover, other logical bugs are removed.

### CODE REVIEW

**Comment by abdullah-siddiqui-git** [comment]
Cycle - 0 review done.

File: app/models/ability.rb [hunk 1]
```diff
@@ -3,32 +3,11 @@
...
-    # The third argument is an optional hash of conditions to further filter the
-    # objects.
-    # For example, here the user can only update published articles.
-    #
-    #   can :update, Article, :published => true
-    #
-    # See the wiki for details:
-    # https://github.com/CanCanCommunity/cancancan/wiki/Defining-Abilities
+  def initialize(user, options = {})
+    return unless user.present?
```
Comment (basil7vals) [inline app/models/ability.rb]: define ability for admin like:
if user.admin?
      can :manage, :all

File: app/views/devise/confirmations/new.html.erb [hunk 2]
```diff
@@ -1,16 +1,24 @@
...
+  <div class="offset-sm-2 col-sm-8">
+    <div class="card">
+      <h1 class="card-header">Resend confirmation instructions</h1>
+      <div class="card-body">
+        <%= form_for(resource, as: resource_name, url: confirmation_path(resource_name), html: { method: :post }) do |f| %>
+          <%= devise_error_messages! %>
 
-<%= form_for(resource, as: resource_name, url: confirmation_path(resource_name), html: { method: :post }) do |f| %>
-  <%= devise_error_messages! %>
+          <div class="field form-group">
```
Comment (basil7vals) [inline app/views/devise/confirmations/new.html.erb]: indentation

File: Gemfile [hunk 3]
```diff
@@ -9,6 +9,8 @@ gem 'devise', '4.1.1'
 gem 'jbuilder', '~> 2.0'
 # Use jquery as the JavaScript library
 gem 'jquery-rails'
+# JQuery turbolinks
+gem 'jquery-turbolinks'
```
Comment (basil7vals) [inline Gemfile]: add versions

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -24,7 +35,12 @@ def current_tenant
   private
 
   def scope_current_tenant
-    if WIHTOUT_SUBDOMAIN_URLS.include?(request.path) && request.subdomain.present?
+    # RENDER 404 RESPONSE IF THE PATH OF THE REQUESTED URL BELONGS TO THE URLS THAT ARE ACCESSIBLE
+    # ONLY WITH SUBDOMAINS AND VICE VERSA
+    # SOME URLS LIKE '/users' ARE ACCESSIBLE WITH BOTH SUBDOMAIN AND WITHOUT SUBDOMAIN.
+    if SUBDOMAIN_INDEPENDENT_URLS.include?([request.method, request.path])
+      Company.current_id = current_tenant.id if request.subdomain.present?
+    elsif WIHTOUT_SUBDOMAIN_URLS.include?(request.path) && request.subdomain.present?
```
Comment (mujtaba7vals) [inline app/controllers/application_controller.rb]: verify

File: app/models/ability.rb [hunk 5]
```diff
@@ -3,32 +3,11 @@
...
+
+    can [:show_change_password_form, :change_password], :member if options[:change_password_member_id] == user.id
+    can [:index, :show], :member
+    can [:new, :create, :privileges, :privileges_show, :edit, :destroy, :update], :member if user.role_id == Role.admin.id
```
Comment (mujtaba7vals) [inline app/models/ability.rb]: go with method flow

File: app/views/devise/sessions/new.html.erb [hunk 6]
```diff
@@ -1,38 +1,41 @@
...
   <div class="offset-sm-2 col-sm-8">
 
-    <h2>Log in</h2>
+    <div class="card">
+      <h1 class="card-header">Log in</h1>
+      <div class="card-body">
+        <%= form_for(resource, as: resource_name, url: session_path(resource_name)) do |f| %>
+          <div class="field form-group">
+            <%= f.label :email %><br />
+            <% if @email.nil?
```
Comment (mujtaba7vals) [inline app/views/devise/sessions/new.html.erb]: one liner implementation

File: app/views/devise/sessions/new.html.erb [hunk 7]
```diff
@@ -1,38 +1,41 @@
...
+        <%= form_for(resource, as: resource_name, url: session_path(resource_name)) do |f| %>
+          <div class="field form-group">
+            <%= f.label :email %><br />
+            <% if @email.nil? 
+              readonly = false
+            else
+              readonly = true
+            end
+            %>
+            <%= f.email_field :email, autofocus: true, class: "form-control form-control-sm", placeholder: "john.doe@gmail.com", value: @email, readonly: readonly %>
```
Comment (mujtaba7vals) [inline app/views/devise/sessions/new.html.erb]: pass `email.present?`  directly

File: app/views/members/index.html.erb [hunk 8]
```diff
@@ -6,14 +6,31 @@
         <em class="text-muted float-right">@<%= @company.name %></em>
       </h1>
       <div class="card-body">
+        <% if can? :privileges, :member %>
+        <p class="text-center text-muted"><em>Want to change roles of members? </em><%= link_to "Manage Privileges", privileges_members_path, class: 'btn btn-sm btn-outline-info' %></p>
```
Comment (mujtaba7vals) [inline app/views/members/index.html.erb]: indentation

File: app/models/ability.rb (hunk 1 above)
Comment (abdullah-siddiqui-git) [reply app/models/ability.rb]: Noted.

File: app/views/devise/confirmations/new.html.erb (hunk 2 above)
Comment (abdullah-siddiqui-git) [reply app/views/devise/confirmations/new.html.erb]: Noted.

File: app/views/devise/sessions/new.html.erb (hunk 6 above)
Comment (basil7vals) [inline app/views/devise/sessions/new.html.erb]: use hidden: true instead

**Comment by basil7vals** [comment]
Cycle - 1 completed

File: Gemfile (hunk 3 above)
Comment (abdullah-siddiqui-git) [reply Gemfile]: Done.

File: app/models/ability.rb (hunk 5 above)
Comment (abdullah-siddiqui-git) [reply app/models/ability.rb]: Done.

File: app/views/devise/sessions/new.html.erb (hunk 7 above)
Comment (abdullah-siddiqui-git) [reply app/views/devise/sessions/new.html.erb]: Resolved.

File: app/views/members/index.html.erb (hunk 8 above)
Comment (abdullah-siddiqui-git) [reply app/views/members/index.html.erb]: Resolved.

File: app/views/devise/sessions/new.html.erb (hunk 6 above)
Comment (abdullah-siddiqui-git) [reply app/views/devise/sessions/new.html.erb]: Done.

File: app/controllers/application_controller.rb (hunk 4 above)
Comment (abdullah-siddiqui-git) [reply app/controllers/application_controller.rb]: Done.

**Pull request merged**